## eBay Car Sales Data Analysis

Magnus Skonberg | January 4th 2022

### Project Intro

The purpose of this project is to utilize basic numpy and pandas library functionality, covered up to this point to perform a practical data analysis. 

In this project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle by user `orgesleka`. While the original dataset isn't available on Kaggle anymore, it can be found [here](https://data.world/data-society/used-cars-data).  The version we'll be working with is a sample of 50k data points that prepared by Dataquest for simulating a less-cleaned version of the data.

The data dictionary, provided with data, is as follows:

* dateCrawled - When this ad was first crawled. All field-values are taken from this date.
* name - Name of the car.
* seller - Whether the seller is private or a dealer.
* offerType - The type of listing
* price - The price on the ad to sell the car.
* abtest - Whether the listing is included in an A/B test.
* vehicleType - The vehicle Type.
* yearOfRegistration - The year in which which year the car was first registered.
* gearbox - The transmission type.
* powerPS - The power of the car in PS.
* model - The car model name.
* kilometer - How many kilometers the car has driven.
* monthOfRegistration - The month in which which year the car was first registered.
* fuelType - What type of fuel the car uses.
* brand - The brand of the car.
* notRepairedDamage - If the car has a damage which is not yet repaired.
* dateCreated - The date on which the eBay listing was created.
* nrOfPictures - The number of pictures in the ad.
* postalCode - The postal code for the location of the vehicle.
* lastSeenOnline - When the crawler saw this ad last online.

The purpose of this project will be to clean and analyze the used car listing data in order to become more familiar with the unique benefits jupyter notebook provides for pandas.

### Load Data

To start we read in our data:

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

What we gauge from above:

* there are 20 columns
* the majority of which are strings
* columns are named in snakecase
* 5 columns have NULL presence

In [4]:
old_column_names = autos.columns

# correct column naming to use lowercase
col_name_mapping = {'dateCrawled' : 'date_crawled',
                    'offerType' : 'offer_type',
                    'abtest' : 'ab_test',
                    'vehicleType' : 'vehicle_type',
                    'yearOfRegistration' : 'registration_year',
                    'powerPS' : 'power_ps',
                    'monthOfRegistration' : 'registration_month',
                    'fuelType' : 'fuel_type',
                    'notRepairedDamage' : 'unrepaired_damage',
                    'dateCreated' : 'ad_created',
                    'nrOfPictures' : 'nr_of_pictures',
                    'postalCode' : 'postal_code',
                    'lastSeen' : 'last_seen'}

autos.rename(columns= col_name_mapping, inplace=True)

# verify output
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

## Explore and Clean

With columns renamed and initial exploration completed, we perform a more in-depth exploration and cleaning of the data:

In [5]:
# Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
# Examples of numeric data stored as text which can be cleaned and converted.

autos.describe(include='all') #drop nr_of_pictures
autos['nr_of_pictures'].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

There are a # of column updates to be made:

* **drop**: seller, offer_type, and nr_of_pictures
* **conversion to numeric**: price, odometer

In [6]:
# drop identified columns
autos = autos.drop(["seller", "offer_type", "nr_of_pictures"], axis=1)

# convert identified columns
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","")
autos["price"] = autos["price"].astype(int)

autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","")
autos["odometer"] = autos["odometer"].astype(int)
autos.rename({"odometer":"odometer_km"}, axis=1, inplace=True)

#verify output
autos.head()


<ipython-input-6-a857ca46bff3>:5: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will*not* be treated as literal strings when regex=True.
  autos["price"] = autos["price"].str.replace("$","")


date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   price  ab_test vehicle_type  registration_year    gearbox  power_ps  \
0   5000  control          bus               2004    manuell       158   
1   8500  control    limousine               1997  automatik       286   
2   8990     test    limousine               2009    manuell       102   
3   4350  control   kleinwagen               2007  automatik        71   
4   1350     test        kombi               2003    manuell         0   

    model  odometer_km  registration_month fuel_type       brand  \
0  andere       150000                   3       lpg     peugeot   
1     7er       150000                   6    benzin         bmw   
2    golf        70000                   7    benzin  volkswagen   
3  fortwo        70000                   6    benzin       smart   
4   focus       150000                   7    benzin        ford   

  unrepaired_damage           ad_created  postal_code            last_seen  
0              nein  2016-03-26 00:00:00        79588  2016-04-06 06:45:54  
1              nein  2016-04-04 00:00:00        71034  2016-04-06 14:45:08  
2              nein  2016-03-26 00:00:00        35394  2016-04-06 20:15:37  
3              nein  2016-03-12 00:00:00        33729  2016-03-15 03:16:28  
4              nein  2016-04-01 00:00:00        39218  2016-04-01 14:38:50

### Explore Odometer and Price

Next up we consider our odometer and price columns to gain a better understanding of ranges and distributions:

In [7]:
# analyze odometer and price columns - look for and then filter out outliers
autos["price"].unique().shape #2357 unique prices

#explore extremes
max(autos["price"]) # vehicles priced at 99999999?
min(autos["price"]) # vehicles priced at 0?

#explore summary stats
autos["price"].describe()

#explore value counts / ranges of interest
autos["price"].value_counts() #0 is the most popular value
autos["price"].value_counts().sort_index(ascending=False).head(20) #most expensive car in world $70mil
autos["price"].value_counts().sort_index(ascending=True).head(20) #most expensive car in world $70mil

#apply filter to price column - filter for $350k and less and $100 and more
autos = autos[autos["price"].between(100,350000)]


In [8]:
# analyze odometer and price columns - look for and then filter out outliers
autos["odometer_km"].unique().shape #13 unique odometer readings

#explore extremes
max(autos["odometer_km"]) # 150000 max km
min(autos["odometer_km"]) # 5000 min km

#explore summary stats
autos["odometer_km"].describe() # majority of vehicles sit in upper km range

#explore value counts / ranges of interest
autos["odometer_km"].value_counts() #150000 is the most popular value

# these values make sense, do not filter

150000    31212
125000     5037
100000     2101
90000      1733
80000      1412
70000      1214
60000      1153
50000      1009
40000       814
30000       777
5000        760
20000       757
10000       245
Name: odometer_km, dtype: int64

I explore two numeric columns (`price` and `odometer_km`). While the `odometer_km` column's data is commonsensical (km range between 10000 and 150000 where we're generally dealing with older vehicles), the `price` range is well below and above reason. We address this by filtering for vehicles priced more than 100 and less than 350k dollars.

### Explore date columns

There are a number of columns with date information, we explore each of these columns to learn more of our listings:

In [9]:
# explore date columns: date_crawled, last_seen, ad_created, registration_month, and registration_year

# elect first 10 characters in the column to extract date
autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False) #.sort_index()
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False) #.sort_index()
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False) #.sort_index()
autos['registration_month'].value_counts(normalize=True, dropna=False) #.sort_index()
autos['registration_year'].describe()

count    48224.000000
mean      2004.730964
std         87.897388
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

From the above date distributions we gather that:
1. `date_crawled`: the site was crawled between March and April of 2016 and %s are relatively evenly distributed
2. `last_seen`: the site was last seen between March and April of 2016 with the greatest concentration in April
3. `ad_created`: ads were created from 2015 to 2016 with the gretest % for more recent (March/April) of dates
4. `registration_month`: the greatest % is for March then 0 (unidentified) while the lowest % is for February
5. `registration_year`: the min is 1000 (before cars were invented) and the max is 9999 (far in the future). There's an outlier presence.

It appears there may be some outlier handling required based on registration month and year columns.

In [10]:
autos['registration_year'].value_counts(normalize=True, dropna=False).tail(10) #.sort_index()

low_autos = autos[autos['registration_year'] < 1900]
low_autos['registration_year'].value_counts(dropna=False) 

1800    2
1001    1
1000    1
1111    1
Name: registration_year, dtype: int64

Based on the above findings we decide to filter out vehicle's whose registration year is before 1900 or after 2016 (the year the data was collected).

In [11]:
# apply filter to price column - filter for $350k and less and $100 and more
autos = autos[autos["registration_year"].between(1900,2016)]

# revisit distribution
autos['registration_year'].value_counts(normalize=True)

2000    0.066966
2005    0.062802
1999    0.062112
2004    0.058228
2003    0.058099
          ...   
1929    0.000022
1931    0.000022
1938    0.000022
1939    0.000022
1952    0.000022
Name: registration_year, Length: 78, dtype: float64

The new distribution appears to make sense. There are 78 different registration years where more recent registered vehicles (ie. 2000) make up the larger % while more distant registered vehicles (ie. 1952) make up a smaller %.

### Explore Price Based on Brand

We utilize aggregation to understand what brands we're dealing with and then the average price based on brand:

In [12]:
# use aggregation to understand the brand column

# identify unique values we want to aggregate by
brand_names = autos['brand'].value_counts().head(20).index.tolist()

# verify list of brand names
#print(brand_names)

I've elected brands based on their presence in the dataset. The top 20 brands in the dataset.

In [13]:
# create empty dictionary to store aggregate data
brand_price = {}

# Loop over unique values, and for each: subset dataframe, calculate mean price, and assign value to dict
for b in brand_names:
    brand_price[b] = autos[autos['brand'] == b]['price'].mean().astype(int)

print(brand_price)

{'volkswagen': 5436, 'bmw': 8381, 'opel': 3005, 'mercedes_benz': 8672, 'audi': 9380, 'ford': 3779, 'renault': 2496, 'peugeot': 3113, 'fiat': 2836, 'seat': 4433, 'skoda': 6409, 'nissan': 4756, 'mazda': 4129, 'smart': 3596, 'citroen': 3796, 'toyota': 5167, 'hyundai': 5411, 'sonstige_autos': 12784, 'volvo': 4993, 'mini': 10639}


The above output provides feedback on pricing based on the brand:
* **highest average price**: sonstige
* **lowest average price**: renault
* **mid price**: everything from ford to volkswagen

There are tiers to this - quite a range of vehicle pricing.

As such, we proceed to explore just the top 6 brands:

In [14]:
# For the top 6 brands: we'll aggregate to understand the average mileage and if there's any visible link with mean price. 

# create empty dictionary to store aggregate data
#brand_pm = {}
brand_mileage = {}

# loop over brands of interest for aggregate price and mileage
#for b in brand_names[0:6]:
#        brand_pm[b] = [ autos[autos['brand'] == b]['price'].mean().astype(int), 
#                        autos[autos['brand'] == b]['odometer_km'].mean().astype(int) ]

#print(brand_pm)

for b in brand_names:
    brand_mileage[b] = autos[autos['brand'] == b]['odometer_km'].mean().astype(int)

# convert dictionaries to series objects
bp_series = pd.Series(brand_price)
bm_series = pd.Series(brand_mileage)

# convert series objects to dataframes
df = pd.DataFrame(bp_series, columns=['mean_price'])
df2 = pd.DataFrame(bm_series, columns=['mean_km'])

# add mileage column to df
df['mean_km'] = df2['mean_km']

print(df.head(6))


               mean_price  mean_km
volkswagen           5436   128799
bmw                  8381   132695
opel                 3005   129384
mercedes_benz        8672   131025
audi                 9380   129245
ford                 3779   124277


We aggregated to understand the average mileage and observe that the correlation between price and kms on the odometer is weak. It's more dependent upon brand.

### Conclusion

To conclude, we cleaned our car data and practiced our familiarity with pandas while determining that brand has the greatest correlation with price for the used cars under our consideration.